### ISSUE TITLE
Support multi-resources with different schemas YAML files

### ISSUE THREAD

**Issue by victornoel** [Azure/vscode-kubernetes-tools#341]
When opening a k8s YAML file, there can be multiple resources descrtibed in it, separated by `---`.

For example:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: app
spec:
  selector:
    matchLabels:
      app: app
  replicas: 1
  template:
    metadata:
      labels:
        app: app
    spec:
      containers:
        - name: app
          image: nginx
      volumes:
        - name: config-volume
          configMap:
            name: app-config
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: app-config
data:
  config: some config
```

In practice, the editor shows an error saying `Matches multiple schemas when only one must validate.` and not validation happens.

I think that we should be able to validate separately the different resources of a yaml file without problems by considering them as if they were different files.

Even though YAML support is implemented with yaml-language-server, according to the discussion in redhat-developer/yaml-language-server#80, the problem should be solved in the present extension and not there.

**Comment by itowlson** [comment]
cc @andxu

**Comment by itowlson** [comment]
@victornoel Just to update, we will take a look at this but the person who knows that code best is tied up on something else.  We're expecting him to come free soon, but if not, I or someone else will pick it up.

**Comment by andxu** [comment]
@victornoel @JPinkney The problems are caused by the differences between json and yaml document: one json file only has one document while one yaml file can have multiple documents, in this case(multiple document), the current behavior for k8s schema provider is to provide anyOf [Deployment, ConfigMap] for the specified yaml, this works well if no validation problems/errors found, but if any errors are founded, for example the second ConfigMap has some validation errors, then the yaml-languange-server will not be able to detect which schema is actually chosen since both [Deployment, ConfigMap] schema is not acceptable, and it will use the first one Deployment to report the errors. The solution to this scenario is to notice the getSchemaForResource may return an Array if multiple documents are detected rather than the tricky anyOf schema to represent multiple documents, I am preparing for the PR.

**Comment by itowlson** [comment]
@victornoel Andy has sent a PR to the upstream YAML extension at https://github.com/redhat-developer/yaml-language-server/pull/81

**Comment by victornoel** [comment]
@itowlson yep, I saw it :) In the end there was maybe no need to open this issue I suppose… I will close it when it is merged then!

**Comment by JPinkney** [comment]
Merged and a new version has been released on our side!

**Comment by victornoel** [comment]
@JPinkney is there something to be modified in vscode-kubernetes-tool to take advantage of that? Because I just tried (0.0.15) and I still see the errors.

**Comment by JPinkney** [comment]
I'm not entirely sure. I just assumed that the fix was correct for this case and that they had it all working when the PR was submitted.

**Comment by victornoel** [comment]
@JPinkney ok, I misunderstood, I thought you were the origin of the PR, so I suppose we will wait for news from @andxu :)

**Comment by victornoel** [comment]
and/or @itowlson

**Comment by itowlson** [comment]
Sorry, we're playing timezone tag a bit here!  @andxu has told me that he will need to make a "very little change" the k8s side after the new vscode-yaml is available in the marketplace.  He will send out a PR.  So this issue remains open for now.

Thanks heaps @JPinkney for the rapid release, and thanks heaps Andy for the upstream fix.

**Comment by itowlson** [comment]
@JPinkney Since the Red Hat YAML extension auto-upgraded, I've started seeing the error toast "Couldn't start client Yaml Support" when starting the k8s extension, with "Error - starting client failed" in the Output window.  This message doesn't _seem_ to originate within the k8s extension, and by the casing of YAML I don't think it can be a system message relating to us trying to activate the YAML extension.  Do you know if these messages could be coming from the YAML extension?

**Comment by itowlson** [comment]
...or there may be something else going on, as now the Dockerfile language server is starting to give me errors...  _sigh_

**Comment by andxu** [comment]
@itowlson the error "Error - starting client failed" is raised from https://github.com/redhat-developer/vscode-yaml/blob/master/src/extension.ts#L29, in vscode-yaml, it seems to use an older way of launching node, so in debug mode, the yaml extension will fail to activate if you are using node lts versoin v8.xx, I think you can change it from:
let debugOptions = { execArgv: ["--nolazy", "--debug=6009"] };
==>
let debugOptions = { execArgv: ["--nolazy", "--inspect"] };
to make you can debug the plugin.

**Comment by itowlson** [comment]
Ah, thanks for the explanation @andxu.

**Comment by itowlson** [comment]
@andxu Hmm, I misunderstood you.  I'm consuming the binary of the RH YAML plug-in from the marketplace, so I can't modify that line.  But I'm not interested in debugging the YAML extension, only the k8s extension.  So I don't really need the language server to run in debug mode anyway.  As a _consumer_ of the YAML extension, what's my workaround?

I'm on Node 10.1 if that makes a difference.

**Comment by itowlson** [comment]
Fixed via #353.

**Comment by victornoel** [comment]
:+1:

thanks!
